The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12884810276/job/35921938055) [(plain)](https://github.com/rust-lang/rust/commit/aeee27b992e1e52bcca795b231af7cad5bbbff24/checks/35921938055/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling build_helper v0.1.0 (/checkout/src/build_helper)
error[E0382]: borrow of moved value: `cc`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:156:33
    |
150 |         let cc = builder.build.cc(target).display().to_string();
    |             -- move occurs because `cc` has type `std::string::String`, which does not implement the `Copy` trait
...
156 |             .map_or_else(|| cc, |ccache| format!("{ccache} {cc}"));
    |                          -- --  ^^^^^^^^                    -- borrow occurs due to use in closure
    |                          |  |   |
    |                          |  variable moved due to use in closure
    |                          value moved into closure here
    |
help: consider cloning the value if the performance cost is acceptable
help: consider cloning the value if the performance cost is acceptable
    |
156 |             .map_or_else(|| cc.clone(), |ccache| format!("{ccache} {cc}"));

error[E0382]: borrow of moved value: `cxx`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:166:38
    |
    |
160 |             let cxx = cxx.display().to_string();
    |                 --- move occurs because `cxx` has type `std::string::String`, which does not implement the `Copy` trait
...
166 |                 .map_or_else(|| cxx, |ccache| format!("{ccache} {cxx}"));
    |                              -- ---  ^^^^^^^^                    --- borrow occurs due to use in closure
    |                              |  |    value borrowed here after move
    |                              |  variable moved due to use in closure
    |                              value moved into closure here
    |
    |
help: consider cloning the value if the performance cost is acceptable
    |
166 |                 .map_or_else(|| cxx.clone(), |ccache| format!("{ccache} {cxx}"));

For more information about this error, try `rustc --explain E0382`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
error: could not compile `bootstrap` (lib) due to 2 previous errors
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 2/5:
##[group]Building bootstrap
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0382]: borrow of moved value: `cc`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:156:33
    |
150 |         let cc = builder.build.cc(target).display().to_string();
    |             -- move occurs because `cc` has type `std::string::String`, which does not implement the `Copy` trait
...
156 |             .map_or_else(|| cc, |ccache| format!("{ccache} {cc}"));
    |                          -- --  ^^^^^^^^                    -- borrow occurs due to use in closure
    |                          |  |   |
    |                          |  variable moved due to use in closure
    |                          value moved into closure here
    |
help: consider cloning the value if the performance cost is acceptable
help: consider cloning the value if the performance cost is acceptable
    |
156 |             .map_or_else(|| cc.clone(), |ccache| format!("{ccache} {cc}"));

error[E0382]: borrow of moved value: `cxx`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:166:38
    |
    |
160 |             let cxx = cxx.display().to_string();
    |                 --- move occurs because `cxx` has type `std::string::String`, which does not implement the `Copy` trait
...
166 |                 .map_or_else(|| cxx, |ccache| format!("{ccache} {cxx}"));
    |                              -- ---  ^^^^^^^^                    --- borrow occurs due to use in closure
    |                              |  |    value borrowed here after move
    |                              |  variable moved due to use in closure
    |                              value moved into closure here
    |
    |
help: consider cloning the value if the performance cost is acceptable
    |
166 |                 .map_or_else(|| cxx.clone(), |ccache| format!("{ccache} {cxx}"));

For more information about this error, try `rustc --explain E0382`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
error: could not compile `bootstrap` (lib) due to 2 previous errors
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 3/5:
##[group]Building bootstrap
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0382]: borrow of moved value: `cc`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:156:33
    |
150 |         let cc = builder.build.cc(target).display().to_string();
    |             -- move occurs because `cc` has type `std::string::String`, which does not implement the `Copy` trait
...
156 |             .map_or_else(|| cc, |ccache| format!("{ccache} {cc}"));
    |                          -- --  ^^^^^^^^                    -- borrow occurs due to use in closure
    |                          |  |   |
    |                          |  variable moved due to use in closure
    |                          value moved into closure here
    |
help: consider cloning the value if the performance cost is acceptable
help: consider cloning the value if the performance cost is acceptable
    |
156 |             .map_or_else(|| cc.clone(), |ccache| format!("{ccache} {cc}"));

error[E0382]: borrow of moved value: `cxx`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:166:38
    |
    |
160 |             let cxx = cxx.display().to_string();
    |                 --- move occurs because `cxx` has type `std::string::String`, which does not implement the `Copy` trait
...
166 |                 .map_or_else(|| cxx, |ccache| format!("{ccache} {cxx}"));
    |                              -- ---  ^^^^^^^^                    --- borrow occurs due to use in closure
    |                              |  |    value borrowed here after move
    |                              |  variable moved due to use in closure
    |                              value moved into closure here
    |
    |
help: consider cloning the value if the performance cost is acceptable
    |
166 |                 .map_or_else(|| cxx.clone(), |ccache| format!("{ccache} {cxx}"));

For more information about this error, try `rustc --explain E0382`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
error: could not compile `bootstrap` (lib) due to 2 previous errors
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 4/5:
##[group]Building bootstrap
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0382]: borrow of moved value: `cc`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:156:33
    |
150 |         let cc = builder.build.cc(target).display().to_string();
    |             -- move occurs because `cc` has type `std::string::String`, which does not implement the `Copy` trait
...
156 |             .map_or_else(|| cc, |ccache| format!("{ccache} {cc}"));
    |                          -- --  ^^^^^^^^                    -- borrow occurs due to use in closure
    |                          |  |   |
    |                          |  variable moved due to use in closure
    |                          value moved into closure here
    |
help: consider cloning the value if the performance cost is acceptable
help: consider cloning the value if the performance cost is acceptable
    |
156 |             .map_or_else(|| cc.clone(), |ccache| format!("{ccache} {cc}"));

error[E0382]: borrow of moved value: `cxx`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:166:38
    |
    |
160 |             let cxx = cxx.display().to_string();
    |                 --- move occurs because `cxx` has type `std::string::String`, which does not implement the `Copy` trait
...
166 |                 .map_or_else(|| cxx, |ccache| format!("{ccache} {cxx}"));
    |                              -- ---  ^^^^^^^^                    --- borrow occurs due to use in closure
    |                              |  |    value borrowed here after move
    |                              |  variable moved due to use in closure
    |                              value moved into closure here
    |
    |
help: consider cloning the value if the performance cost is acceptable
    |
166 |                 .map_or_else(|| cxx.clone(), |ccache| format!("{ccache} {cxx}"));

For more information about this error, try `rustc --explain E0382`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
error: could not compile `bootstrap` (lib) due to 2 previous errors
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 5/5:
##[group]Building bootstrap
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0382]: borrow of moved value: `cc`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:156:33
    |
150 |         let cc = builder.build.cc(target).display().to_string();
    |             -- move occurs because `cc` has type `std::string::String`, which does not implement the `Copy` trait
...
156 |             .map_or_else(|| cc, |ccache| format!("{ccache} {cc}"));
    |                          -- --  ^^^^^^^^                    -- borrow occurs due to use in closure
    |                          |  |   |
    |                          |  variable moved due to use in closure
    |                          value moved into closure here
    |
help: consider cloning the value if the performance cost is acceptable
help: consider cloning the value if the performance cost is acceptable
    |
156 |             .map_or_else(|| cc.clone(), |ccache| format!("{ccache} {cc}"));

error[E0382]: borrow of moved value: `cxx`
##[error]   --> src/bootstrap/src/core/build_steps/gcc.rs:166:38
    |
    |
160 |             let cxx = cxx.display().to_string();
    |                 --- move occurs because `cxx` has type `std::string::String`, which does not implement the `Copy` trait
...
166 |                 .map_or_else(|| cxx, |ccache| format!("{ccache} {cxx}"));
    |                              -- ---  ^^^^^^^^                    --- borrow occurs due to use in closure
    |                              |  |    value borrowed here after move
    |                              |  variable moved due to use in closure
    |                              value moved into closure here
    |
    |
help: consider cloning the value if the performance cost is acceptable
    |
166 |                 .map_or_else(|| cxx.clone(), |ccache| format!("{ccache} {cxx}"));

For more information about this error, try `rustc --explain E0382`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
error: could not compile `bootstrap` (lib) due to 2 previous errors
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked
make: *** [Makefile:98: prepare] Error 1
The command has failed after 5 attempts.
  local time: Tue Jan 21 10:22:41 UTC 2025
  network time: Tue, 21 Jan 2025 10:22:41 GMT
```

</details>

--- a/src/bootstrap/src/core/build_steps/gcc.rs
+++ b/src/bootstrap/src/core/build_steps/gcc.rs
@@ -12,6 +12,8 @@ use std::fs;
 use std::path::PathBuf;
 use std::sync::OnceLock;
 
+use build_helper::ci::CiEnv;
+
 use crate::Kind;
 use crate::core::builder::{Builder, RunConfig, ShouldRun, Step};
 use crate::core::config::TargetSelection;
@@ -112,16 +114,60 @@ impl Step for Gcc {
             return true;
         }
 
-        command(root.join("contrib/download_prerequisites")).current_dir(&root).run(builder);
-        command(root.join("configure"))
+        // GCC creates files (e.g. symlinks to the downloaded dependencies)
+        // in the source directory, which does not work with our CI setup, where we mount
+        // source directories as read-only on Linux.
+        // Therefore, as a part of the build in CI, we first copy the whole source directory
+        // to the build directory, and perform the build from there.
+        let src_dir = if CiEnv::is_ci() {
+            let src_dir = builder.gcc_out(target).join("src");
+            if src_dir.exists() {
+                builder.remove_dir(&src_dir);
+            }
+            builder.create_dir(&src_dir);
+            builder.cp_link_r(&root, &src_dir);
+            src_dir
+        } else {
+            root
+        };
+
+        command(src_dir.join("contrib/download_prerequisites")).current_dir(&src_dir).run(builder);
+        let mut configure_cmd = command(src_dir.join("configure"));
+        configure_cmd
             .current_dir(&out_dir)
+            // On CI, we compile GCC with Clang.
+            // The -Wno-everything flag is needed to make GCC compile with Clang 19.
+            // `-g -O2` are the default flags that are otherwise used by Make.
+            // FIXME(kobzol): change the flags once we have [gcc] configuration in config.toml.
+            .env("CXXFLAGS", "-Wno-everything -g -O2")
+            .env("CFLAGS", "-Wno-everything -g -O2")
             .arg("--enable-host-shared")
             .arg("--enable-languages=jit")
             .arg("--enable-checking=release")
             .arg("--disable-bootstrap")
             .arg("--disable-multilib")
-            .arg(format!("--prefix={}", install_dir.display()))
-            .run(builder);
+            .arg(format!("--prefix={}", install_dir.display()));
+        let cc = builder.build.cc(target).display().to_string();
+        let cc = builder
+            .build
+            .config
+            .ccache
+            .as_ref()
+            .map_or_else(|| cc.clone(), |ccache| format!("{ccache} {cc}"));
+        configure_cmd.env("CC", cc);
+
+        if let Ok(ref cxx) = builder.build.cxx(target) {
+            let cxx = cxx.display().to_string();
+            let cxx = builder
+                .build
+                .config
+                .ccache
+                .as_ref()
+                .map_or_else(|| cxx.clone(), |ccache| format!("{ccache} {cxx}"));
+            configure_cmd.env("CXX", cxx);
+        }
+        configure_cmd.run(builder);
+
         command("make").current_dir(&out_dir).arg(format!("-j{}", builder.jobs())).run(builder);
         command("make").current_dir(&out_dir).arg("install").run(builder);
 

--- a/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile
+++ b/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile
@@ -34,6 +34,7 @@ RUN yum upgrade -y && \
       python3 \
       unzip \
       wget \
+      flex \
       xz \
       zlib-devel.i686 \
       zlib-devel.x86_64 \

--- a/src/ci/docker/scripts/build-zstd.sh
+++ b/src/ci/docker/scripts/build-zstd.sh
@@ -25,5 +25,11 @@ cd zstd-$ZSTD
 CFLAGS=-fPIC hide_output make -j$(nproc) VERBOSE=1
 hide_output make install
 
+# It doesn't seem to be possible to move destination directory
+# of the `make install` above. We thus copy the built artifacts
+# manually to our custom rustroot, so that it can be found through
+# LD_LIBRARY_PATH.
+cp /usr/local/lib/libzstd* /rustroot/lib64
+
 cd ..
 rm -rf zstd-$ZSTD

